### ISSUE TITLE
CORS eats custom board backgrounds

### ISSUE THREAD

**Issue by Archenoth** [WandererXII/lishogi#675]
So! When you set a custom board image, it loads nicely on most pages!
![board with the background loaded sucessfully](https://github.com/WandererXII/lishogi/assets/1385527/7ebf7c89-c4b5-4ca3-83d7-30938c12320d)
([Image source](https://twitter.com/suger502/status/1647544864387604481), plus a [direct url link here](https://pbs.twimg.com/media/Ft0PzuiakAAcXiF?format=jpg&name=orig) if you wanted to test it)

However, on Puzzles, User Analysis, the normal Analyze page, Practice, Relay, and Studies, the board instead looks like this
![board without the background loaded, also my embarrassingly low puzzle score is visible here (it's 1164)](https://github.com/WandererXII/lishogi/assets/1385527/372feff7-4459-4ae0-8393-69a55574d672)

And on these pages, this message shows up in the console:
![the CORS error in the console](https://github.com/WandererXII/lishogi/assets/1385527/5674eab9-3d2c-48a6-81a2-2e7db8accb2d)

I've narrowed it down to the server replying with "`Cross-Origin-Opener-Policy: same-origin`" when looking at a lot of pages (the broken ones)
![response headers for a request to the /training/daily page, with the Cross-Origin-Opener-Policy being set to "same-origin" highlighted for emphasis](https://github.com/WandererXII/lishogi/assets/1385527/12405d35-a5d8-4435-a59c-69b368f4ef99)

...but not all! (the working ones)
![response headers for a request to a page where i'm playing an ai, which doesn't have Cross-Origin-Opener-Policy set](https://github.com/WandererXII/lishogi/assets/1385527/11437c93-df5a-40db-9041-2ab06863ad1f)

These headers are added in this method!
https://github.com/WandererXII/lishogi/blob/2bd070c3ae2b18fee5173e462d51cb906ccc8233/app/controllers/LilaController.scala#L69-L74

Which [is called when loading all of the pages that break custom board URLs](https://github.com/search?q=repo%3AWandererXII/lishogi%20EnableSharedArrayBuffer&type=code):
![a reference search for the LilaController#EnableSharedArrayBuffer method, which shows it is called in all of the pages previously mentioned that were broken](https://github.com/WandererXII/lishogi/assets/1385527/1a73de4f-0a07-42d7-b518-42234bb15829)

It looks like the original reason this header was added was [to enable multi-threaded WASM in firefox](https://github.com/lichess-org/lila/issues/5784), though I'm not 100% sure how this is related, or if there is a workaround

Is there a way that custom board URLs could be used on all pages across the interface..? (And is there anything I could do to help if there is?)

**Comment by WandererXII** [comment]
Hey! Thanks for submitting the issues.

To get WASM engines to work we need to have:
`Cross-Origin-Embedder-Policy "require-corp"`.
This unfortunately means that all loaded resources need to have: `Cross-Origin-Resource-Policy: cross-origin`.

Chrome supports `credentialless` instead of  `require-corp` header (more about that here https://developer.chrome.com/blog/coep-credentialless-origin-trial/), which eliminates that need, so with chrome it should work.

It seems firefox supports it as an origin trial - https://github.com/mozilla/standards-positions/issues/539#issuecomment-1224320026. But lishogi wouldn't recognize that anyway, since we only check for chrome and there were/are some issues with it? Might need to look into it.

I might set up a proxy on heroku later tomorrow as a dirty solution. Another option would be to host all the images on lishogi...

In short, the image needs to have this header: `Cross-Origin-Resource-Policy: cross-origin`.

**Comment by WandererXII** [comment]
Deployed this: https://github.com/WandererXII/cors-anywhere to render.com.

Simply append the URL with your picture to `https://lishogi-proxy.onrender.com/`

For your image replace the link you use with this: `https://lishogi-proxy.onrender.com/https://pbs.twimg.com/media/Ft0PzuiakAAcXiF?format=jpg&name=orig` and it should work.

I can't guarantee that render.com will keep hosting the proxy, but for now it should work.

**Comment by Archenoth** [comment]
Oh dang! Thank you! I can confirm this works!

Though, if changing CORS stuff is hard, another thing that I sometimes do to get around things like this is to copy the base64 URI of an image so that it doesn't need to be hosted from a third party
![the right click context menu of an image in the firefox web inspector copy section with the option to copy the "Image Data-Url" highlighted](https://github.com/WandererXII/lishogi/assets/1385527/b5f4af96-abbf-4c09-96d0-32aff3674dec)

The benefit of that is that images can't be hotlinked by third party sites, since the img `src` is just the image data itself, and it works without any additional setup

However, the downside of that is that those URIs can also be really long, so you'd want to enforce a reasonable length per user. Also, the TS rejects any URIs that start with anything besides `https://` or `//`:
https://github.com/WandererXII/lishogi/blob/2bd070c3ae2b18fee5173e462d51cb906ccc8233/ui/dasher/src/util.ts#L31-L34

I'm not saying you should do this or anything, but I wanted to bring it up in case you didn't want to maintain a separate proxy altogether~ (Also, thanks again for setting that up! I really do appreciate it!)

**Comment by WandererXII** [comment]
> Oh dang! Thank you! I can confirm this works!
> 
> Though, if changing CORS stuff is hard, another thing that I sometimes do to get around things like this is to copy the base64 URI of an image so that it doesn't need to be hosted from a third party

> The benefit of that is that images can't be hotlinked by third party sites, since the img `src` is just the image data itself, and it works without any additional setup
> 
> However, the downside of that is that those URIs can also be really long, so you'd want to enforce a reasonable length per user. Also, the TS rejects any URIs that start with anything besides `https://` or `//`:
> 
Glad it works!

Allowing the data-url could work, but the data-url would have to be sent from the server with every html doc, which would be rather inefficient. Also storing big data-urls in the _pref_ collection in db could deteriorate performance/memory usage.

I considered to host the images myself on a separate server, providing _upload an image_ button to users, rather than _paste a link_. But for now I don't wanna get into that.
The current implementation works for chrome based browsers (https://caniuse.com/mdn-http_headers_cross-origin-embedder-policy_credentialless), the issue is firefox and safari. Hopefully firefox will soon start suporting `credentialless`.

**Comment by Archenoth** [comment]
This makes sense! And thanks again for the help..!

Since there really isn't anything else that can be done, I think I'll close this issue for now~
